# Goal: To Clean and Analyze Used Car Listings

In [106]:
import pandas as pd
import numpy as np

## 1. Exploring Data

In [107]:
autos = pd.read_csv("autos.csv", encoding = "Latin-1")

In [108]:
autos.head(3)

dateCrawled                                        name  seller  \
0  2016-03-26 17:47:46            Peugeot_807_160_NAVTECH_ON_BOARD  privat   
1  2016-04-04 13:38:56  BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik  privat   
2  2016-03-26 18:57:24                  Volkswagen_Golf_1.6_United  privat   

  offerType   price   abtest vehicleType  yearOfRegistration    gearbox  \
0   Angebot  $5,000  control         bus                2004    manuell   
1   Angebot  $8,500  control   limousine                1997  automatik   
2   Angebot  $8,990     test   limousine                2009    manuell   

   powerPS   model   odometer  monthOfRegistration fuelType       brand  \
0      158  andere  150,000km                    3      lpg     peugeot   
1      286     7er  150,000km                    6   benzin         bmw   
2      102    golf   70,000km                    7   benzin  volkswagen   

  notRepairedDamage          dateCreated  nrOfPictures  postalCode  \
0              nein  2016-03-26 00:00:00             0       79588   
1              nein  2016-04-04 00:00:00             0       71034   
2              nein  2016-03-26 00:00:00             0       35394   

              lastSeen  
0  2016-04-06 06:45:54  
1  2016-04-06 14:45:08  
2  2016-04-06 20:15:37

In [109]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   dateCrawled          50000 non-null  object
 1   name                 50000 non-null  object
 2   seller               50000 non-null  object
 3   offerType            50000 non-null  object
 4   price                50000 non-null  object
 5   abtest               50000 non-null  object
 6   vehicleType          44905 non-null  object
 7   yearOfRegistration   50000 non-null  int64 
 8   gearbox              47320 non-null  object
 9   powerPS              50000 non-null  int64 
 10  model                47242 non-null  object
 11  odometer             50000 non-null  object
 12  monthOfRegistration  50000 non-null  int64 
 13  fuelType             45518 non-null  object
 14  brand                50000 non-null  object
 15  notRepairedDamage    40171 non-null  object
 16  date

In [110]:
autos.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

## 2. Cleaning Column Names

In [111]:
autos.rename({'yearOfRegistration':'registration_year',
              'monthOfRegistration':'registration_month',
              'notRepairedDamage':'unrepaired_damage',
              'dateCreated':'ad_created',
              'dateCrawled':'date_crawled',
              'offerType':'offer_type',
              'price':'price_in_dollars',
              'fuelType':'fuel_type',
              'notRepairedDamage':'not_repaired_damage',
              'nrOfPictures':'nr_pictures',
              'postalCode':'postal_code',
              'lastSeen':'last_seen',
              'vehicleType':'vehicle_type'}, axis = 1, inplace = True)

In [112]:
autos.head(3)

date_crawled                                        name  seller  \
0  2016-03-26 17:47:46            Peugeot_807_160_NAVTECH_ON_BOARD  privat   
1  2016-04-04 13:38:56  BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik  privat   
2  2016-03-26 18:57:24                  Volkswagen_Golf_1.6_United  privat   

  offer_type price_in_dollars   abtest vehicle_type  registration_year  \
0    Angebot           $5,000  control          bus               2004   
1    Angebot           $8,500  control    limousine               1997   
2    Angebot           $8,990     test    limousine               2009   

     gearbox  powerPS   model   odometer  registration_month fuel_type  \
0    manuell      158  andere  150,000km                   3       lpg   
1  automatik      286     7er  150,000km                   6    benzin   
2    manuell      102    golf   70,000km                   7    benzin   

        brand not_repaired_damage           ad_created  nr_pictures  \
0     peugeot                nein  2016-03-26 00:00:00            0   
1         bmw                nein  2016-04-04 00:00:00            0   
2  volkswagen                nein  2016-03-26 00:00:00            0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37

## 3. Exploring Columns 

In [113]:
autos.describe(include='all')

date_crawled         name  seller offer_type price_in_dollars  \
count                 50000        50000   50000      50000            50000   
unique                48213        38754       2          2             2357   
top     2016-04-04 16:40:33  Ford_Fiesta  privat    Angebot               $0   
freq                      3           78   49999      49999             1421   
mean                    NaN          NaN     NaN        NaN              NaN   
std                     NaN          NaN     NaN        NaN              NaN   
min                     NaN          NaN     NaN        NaN              NaN   
25%                     NaN          NaN     NaN        NaN              NaN   
50%                     NaN          NaN     NaN        NaN              NaN   
75%                     NaN          NaN     NaN        NaN              NaN   
max                     NaN          NaN     NaN        NaN              NaN   

       abtest vehicle_type  registration_year  gearbox       powerPS  model  \
count   50000        44905       50000.000000    47320  50000.000000  47242   
unique      2            8                NaN        2           NaN    245   
top      test    limousine                NaN  manuell           NaN   golf   
freq    25756        12859                NaN    36993           NaN   4024   
mean      NaN          NaN        2005.073280      NaN    116.355920    NaN   
std       NaN          NaN         105.712813      NaN    209.216627    NaN   
min       NaN          NaN        1000.000000      NaN      0.000000    NaN   
25%       NaN          NaN        1999.000000      NaN     70.000000    NaN   
50%       NaN          NaN        2003.000000      NaN    105.000000    NaN   
75%       NaN          NaN        2008.000000      NaN    150.000000    NaN   
max       NaN          NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuel_type       brand  \
count       50000        50000.000000     45518       50000   
unique         13                 NaN         7          40   
top     150,000km                 NaN    benzin  volkswagen   
freq        32424                 NaN     30107       10687   
mean          NaN            5.723360       NaN         NaN   
std           NaN            3.711984       NaN         NaN   
min           NaN            0.000000       NaN         NaN   
25%           NaN            3.000000       NaN         NaN   
50%           NaN            6.000000       NaN         NaN   
75%           NaN            9.000000       NaN         NaN   
max           NaN           12.000000       NaN         NaN   

       not_repaired_damage           ad_created  nr_pictures   postal_code  \
count                40171                50000      50000.0  50000.000000   
unique                   2                   76          NaN           NaN   
top                   nein  2016-04-03 00:00:00          NaN           NaN   
freq                 35232                 1946          NaN           NaN   
mean                   NaN                  NaN          0.0  50813.627300   
std                    NaN                  NaN          0.0  25779.747957   
min                    NaN                  NaN          0.0   1067.000000   
25%                    NaN                  NaN          0.0  30451.000000   
50%                    NaN                  NaN          0.0  49577.000000   
75%                    NaN                  NaN          0.0  71540.000000   
max                    NaN                  NaN          0.0  99998.000000   

                  last_seen  
count                 50000  
unique                39481  
top     2016-04-07 06:17:27  
freq                      8  
mean                    NaN  
std                     NaN  
min                     NaN  
25%                     NaN  
50%                     NaN  
75%                     NaN  
max                     NaN

In [114]:
autos["seller"].value_counts()

privat        49999
gewerblich        1
Name: seller, dtype: int64

In [115]:
autos["offer_type"].value_counts()

Angebot    49999
Gesuch         1
Name: offer_type, dtype: int64

In [116]:
autos["abtest"].value_counts()

test       25756
control    24244
Name: abtest, dtype: int64

In [117]:
autos["gearbox"].value_counts()

manuell      36993
automatik    10327
Name: gearbox, dtype: int64

In [118]:
autos["not_repaired_damage"].value_counts()

nein    35232
ja       4939
Name: not_repaired_damage, dtype: int64

In [119]:
autos["nr_pictures"].value_counts()

0    50000
Name: nr_pictures, dtype: int64

### Observations

- Columns like `seller`,`offer type` & `nr pictures` can be dropped as they have only one value or 0 values in them and aren't very useful to analyze

- Columns like `abtest` and `gearbox` can be investigated more to find out further information

- Any examples of numeric data stored as text that needs to be cleaned:

    `date_crawled`
    `price`
    `odometer`
    `ad_created`
    `last_seen`

In [120]:
autos["price_in_dollars"].unique()

array(['$5,000', '$8,500', '$8,990', ..., '$385', '$22,200', '$16,995'],
      dtype=object)

In [121]:
# Changing the price_in_dollar column to int
autos["price_in_dollars"] = autos["price_in_dollars"].str.replace("$","")
autos["price_in_dollars"] = autos["price_in_dollars"].str.replace(",","")
autos["price_in_dollars"] = autos["price_in_dollars"].astype(int)

In [122]:
autos["price_in_dollars"].unique()

array([ 5000,  8500,  8990, ...,   385, 22200, 16995])

In [123]:
#Changing odometer column to int and renaming
autos.rename({"odometer":"odometer_km"}, axis=1, inplace=True)
autos["odometer_km"] = autos["odometer_km"].str.replace("kms","")
autos["odometer_km"] = autos["odometer_km"].str.replace("km","")
autos["odometer_km"] = autos["odometer_km"].str.replace(",","")
autos["odometer_km"] = autos["odometer_km"].astype(int)

In [124]:
autos["odometer_km"].unique()

array([150000,  70000,  50000,  80000,  10000,  30000, 125000,  90000,
        20000,  60000,   5000, 100000,  40000])

### Removing the outliers from both these columns

### Analyzing and removing outliers from the Price column

In [125]:
autos["price_in_dollars"].describe().round()

count       50000.0
mean         9840.0
std        481104.0
min             0.0
25%          1100.0
50%          2950.0
75%          7200.0
max      99999999.0
Name: price_in_dollars, dtype: float64

In [126]:
autos["price_in_dollars"].value_counts().sort_index(ascending=False).head(15)

99999999    1
27322222    1
12345678    3
11111111    2
10000000    1
3890000     1
1300000     1
1234566     1
999999      2
999990      1
350000      1
345000      1
299000      1
295000      1
265000      1
Name: price_in_dollars, dtype: int64

Let's see to which vehicle the value of 265000 corresponds. 

In the event that this price is consistent with the vehicle, we can take it as a reference.

In [127]:
autos[autos["price_in_dollars"] == 265000]

date_crawled                                               name  \
12682  2016-03-28 22:48:01  Porsche_GT3_RS__PCCB__Lift___grosser_Exklusiv_...   

       seller offer_type  price_in_dollars   abtest vehicle_type  \
12682  privat    Angebot            265000  control        coupe   

       registration_year    gearbox  powerPS model  odometer_km  \
12682               2016  automatik      500   911         5000   

       registration_month fuel_type    brand not_repaired_damage  \
12682                   3    benzin  porsche                nein   

                ad_created  nr_pictures  postal_code            last_seen  
12682  2016-03-28 00:00:00            0        70193  2016-04-05 03:44:51

This model of Porsche does have a similar price range. So let's look at the cars above this price

In [128]:
top_list = autos[autos["price_in_dollars"] >= 265000].sort_index(ascending= False)
top_list.head(1)

date_crawled         name  seller offer_type  price_in_dollars  \
47634  2016-04-04 21:25:21  Ferrari_FXX  privat    Angebot           3890000   

      abtest vehicle_type  registration_year gearbox  powerPS model  \
47634   test        coupe               2006     NaN      799   NaN   

       odometer_km  registration_month fuel_type           brand  \
47634         5000                   7       NaN  sonstige_autos   

      not_repaired_damage           ad_created  nr_pictures  postal_code  \
47634                nein  2016-04-04 00:00:00            0        60313   

                 last_seen  
47634  2016-04-05 12:07:37

In [129]:
#Selecting only the required columns
top_list[["price_in_dollars","brand","model","vehicle_type"]].sort_index(ascending=False)

price_in_dollars           brand        model vehicle_type
47634           3890000  sonstige_autos          NaN        coupe
47598          12345678            opel       vectra    limousine
43049            999999      volkswagen  transporter          bus
42221          27322222         citroen           c4    limousine
39705          99999999   mercedes_benz     s_klasse    limousine
39377          12345678           volvo          v40          NaN
37585            999990      volkswagen        jetta    limousine
36818            350000         porsche          911        coupe
35923            295000         porsche          911       cabrio
34723            299000         porsche          911        coupe
27371          12345678            fiat        punto          NaN
24384          11111111      volkswagen          NaN          NaN
22947           1234566             bmw          NaN        kombi
14715            345000  sonstige_autos          NaN       cabrio
12682            265000         porsche          911        coupe
11137          10000000  sonstige_autos          NaN        coupe
7814            1300000  sonstige_autos          NaN        coupe
2897           11111111            ford       escort    limousine
514              999999            ford        focus        kombi

- There are some strange values like 1234566 111111 12345678 which do not seem to be appropriate
- Quite a few cars belong to the `sonstige_autos` category

In [130]:
top_list[top_list['brand'] == 'sonstige_autos']

date_crawled                                               name  \
47634  2016-04-04 21:25:21                                        Ferrari_FXX   
14715  2016-03-30 08:37:24                 Rolls_Royce_Phantom_Drophead_Coupe   
11137  2016-03-29 23:52:57  suche_maserati_3200_gt_Zustand_unwichtig_laufe...   
7814   2016-04-04 11:53:31                                        Ferrari_F40   

       seller offer_type  price_in_dollars   abtest vehicle_type  \
47634  privat    Angebot           3890000     test        coupe   
14715  privat    Angebot            345000  control       cabrio   
11137  privat    Angebot          10000000  control        coupe   
7814   privat    Angebot           1300000  control        coupe   

       registration_year    gearbox  powerPS model  odometer_km  \
47634               2006        NaN      799   NaN         5000   
14715               2012  automatik      460   NaN        20000   
11137               1960    manuell      368   NaN       100000   
7814                1992        NaN        0   NaN        50000   

       registration_month fuel_type           brand not_repaired_damage  \
47634                   7       NaN  sonstige_autos                nein   
14715                   8    benzin  sonstige_autos                nein   
11137                   1    benzin  sonstige_autos                nein   
7814                   12       NaN  sonstige_autos                nein   

                ad_created  nr_pictures  postal_code            last_seen  
47634  2016-04-04 00:00:00            0        60313  2016-04-05 12:07:37  
14715  2016-03-30 00:00:00            0        73525  2016-04-07 00:16:26  
11137  2016-03-29 00:00:00            0        73033  2016-04-06 21:18:11  
7814   2016-04-04 00:00:00            0        60598  2016-04-05 11:34:11

Since these values are justified for their general prices, we can remove everything other than the 'sonstige_autos' and 'Porsche'

In [131]:
autos.iloc[11137,3] = 507230

In [132]:
remove_bad_price_bool = ((top_list["brand"] != 'sonstige_autos') & (top_list["brand"] != 'porsche'))
remove_bad_price_bool.head(3) # sample of the boolean list to be use.

47634    False
47598     True
43049     True
Name: brand, dtype: bool

In [133]:
bad_cars = top_list[remove_bad_price_bool] # type dataframe! This is the dataframe to remove from check_price

In [134]:
len(bad_cars)

11

In [135]:
bad_cars #List of the outlier cars we are removing from the top

date_crawled                                               name  \
47598  2016-03-31 18:56:54  Opel_Vectra_B_1_6i_16V_Facelift_Tuning_Showcar...   
43049  2016-03-21 19:53:52                                      2_VW_Busse_T3   
42221  2016-03-08 20:39:05                                  Leasinguebernahme   
39705  2016-03-22 14:58:27                        Tausch_gegen_gleichwertiges   
39377  2016-03-08 23:53:51                        Tausche_volvo_v40_gegen_van   
37585  2016-03-29 11:38:54                                Volkswagen_Jetta_GT   
27371  2016-03-09 15:45:47                                         Fiat_Punto   
24384  2016-03-21 13:57:51                            Schlachte_Golf_3_gt_tdi   
22947  2016-03-22 12:54:19                         Bmw_530d_zum_ausschlachten   
2897   2016-03-12 21:50:57   Escort_MK_1_Hundeknochen_zum_umbauen_auf_RS_2000   
514    2016-03-17 09:53:08                   Ford_Focus_Turnier_1.6_16V_Style   

       seller offer_type  price_in_dollars   abtest vehicle_type  \
47598  privat    Angebot          12345678  control    limousine   
43049  privat    Angebot            999999     test          bus   
42221  privat    Angebot          27322222  control    limousine   
39705  privat    Angebot          99999999  control    limousine   
39377  privat    Angebot          12345678  control          NaN   
37585  privat    Angebot            999990     test    limousine   
27371  privat    Angebot          12345678  control          NaN   
24384  privat    Angebot          11111111     test          NaN   
22947  privat    Angebot           1234566  control        kombi   
2897   privat    Angebot          11111111     test    limousine   
514    privat    Angebot            999999     test        kombi   

       registration_year    gearbox  powerPS        model  odometer_km  \
47598               2001    manuell      101       vectra       150000   
43049               1981    manuell       70  transporter       150000   
42221               2014    manuell      163           c4        40000   
39705               1999  automatik      224     s_klasse       150000   
39377               2018    manuell       95          v40       150000   
37585               1985    manuell      111        jetta       150000   
27371               2017        NaN       95        punto       150000   
24384               1995        NaN        0          NaN       150000   
22947               1999  automatik      190          NaN       150000   
2897                1973    manuell       48       escort        50000   
514                 2009    manuell      101        focus       125000   

       registration_month fuel_type          brand not_repaired_damage  \
47598                   3    benzin           opel                nein   
43049                   1    benzin     volkswagen                 NaN   
42221                   2    diesel        citroen                 NaN   
39705                   9    benzin  mercedes_benz                 NaN   
39377                   6       NaN          volvo                nein   
37585                  12    benzin     volkswagen                  ja   
27371                   0       NaN           fiat                 NaN   
24384                   0       NaN     volkswagen                 NaN   
22947                   2    diesel            bmw                 NaN   
2897                    3    benzin           ford                nein   
514                     4    benzin           ford                nein   

                ad_created  nr_pictures  postal_code            last_seen  
47598  2016-03-31 00:00:00            0         4356  2016-03-31 18:56:54  
43049  2016-03-21 00:00:00            0        99880  2016-03-28 17:18:28  
42221  2016-03-08 00:00:00            0        76532  2016-03-08 20:39:05  
39705  2016-03-22 00:00:00            0        73525  2016-04-06 05:15:30  
39377  2016-03-08 00:00:00            0        14542  2016-04-06 2

In [136]:
for x in bad_cars.index:
    autos.drop([x], inplace=True)

In [137]:
autos.describe().round()

price_in_dollars  registration_year  powerPS  odometer_km  \
count           49989.0            49989.0  49989.0      49989.0   
mean             6025.0             2005.0    116.0     125732.0   
std             49134.0              106.0    209.0      40042.0   
min                 0.0             1000.0      0.0       5000.0   
25%              1100.0             1999.0     70.0     125000.0   
50%              2950.0             2003.0    105.0     150000.0   
75%              7200.0             2008.0    150.0     150000.0   
max          10000000.0             9999.0  17700.0     150000.0   

       registration_month  nr_pictures  postal_code  
count             49989.0      49989.0      49989.0  
mean                  6.0          0.0      50814.0  
std                   4.0          0.0      25777.0  
min                   0.0          0.0       1067.0  
25%                   3.0          0.0      30451.0  
50%                   6.0          0.0      49577.0  
75%                   9.0          0.0      71522.0  
max                  12.0          0.0      99998.0

In [138]:
#Now assesing the bottom of the list
autos["price_in_dollars"].value_counts().sort_index(ascending=True).head(15)

0     1421
1      156
2        3
3        1
5        2
8        1
9        1
10       7
11       2
12       3
13       2
14       1
15       2
17       3
18       1
Name: price_in_dollars, dtype: int64

In [139]:
#Lets check on the sum of first quartile 
autos[autos["price_in_dollars"].between(0,1100)].loc[:,"price_in_dollars"].sum()

6976424

In [140]:
#Since this is a big range, lets check ranges inside this quartile
perct_75 = autos[autos["price_in_dollars"].between(850,1100)].loc[:,"price_in_dollars"].sum()
perct_50 = autos[autos["price_in_dollars"].between(599,850)].loc[:,"price_in_dollars"].sum()
perct_25 = autos[autos["price_in_dollars"].between(300,599)].loc[:,"price_in_dollars"].sum()

In [141]:
perct_75, perct_50, perct_25

(3227534, 2462601, 1431649)

Since only the value after the last 25 percent of the 11st quartile is a huge value, we take only from 850 USD

In [142]:
autos = autos[autos["price_in_dollars"].between(850,10000000)]

In [143]:
autos["odometer_km"].describe().round()

count     40786.0
mean     123695.0
std       40347.0
min        5000.0
25%      100000.0
50%      150000.0
75%      150000.0
max      150000.0
Name: odometer_km, dtype: float64

### Exploring Date Columns

In [144]:
autos[['date_crawled','ad_created','last_seen']][0:5]

date_crawled           ad_created            last_seen
0  2016-03-26 17:47:46  2016-03-26 00:00:00  2016-04-06 06:45:54
1  2016-04-04 13:38:56  2016-04-04 00:00:00  2016-04-06 14:45:08
2  2016-03-26 18:57:24  2016-03-26 00:00:00  2016-04-06 20:15:37
3  2016-03-12 16:58:10  2016-03-12 00:00:00  2016-03-15 03:16:28
4  2016-04-01 14:38:50  2016-04-01 00:00:00  2016-04-01 14:38:50

In [145]:
autos['date_crawled'].str[:10].value_counts(normalize=True, dropna=False).sort_index()

2016-03-05    0.025573
2016-03-06    0.014147
2016-03-07    0.035600
2016-03-08    0.032683
2016-03-09    0.032462
2016-03-10    0.033075
2016-03-11    0.032658
2016-03-12    0.037415
2016-03-13    0.016035
2016-03-14    0.036434
2016-03-15    0.033639
2016-03-16    0.029177
2016-03-17    0.030770
2016-03-18    0.012921
2016-03-19    0.035110
2016-03-20    0.038028
2016-03-21    0.037439
2016-03-22    0.032805
2016-03-23    0.032315
2016-03-24    0.028956
2016-03-25    0.030721
2016-03-26    0.032879
2016-03-27    0.031310
2016-03-28    0.035208
2016-03-29    0.033713
2016-03-30    0.033100
2016-03-31    0.031261
2016-04-01    0.034424
2016-04-02    0.036238
2016-04-03    0.039156
2016-04-04    0.036802
2016-04-05    0.013264
2016-04-06    0.003212
2016-04-07    0.001471
Name: date_crawled, dtype: float64

In [146]:
autos['ad_created'].str[:10].value_counts(normalize=True, dropna=False).sort_index()

2015-06-11    0.000025
2015-08-10    0.000025
2015-09-09    0.000025
2015-11-10    0.000025
2015-12-05    0.000025
                ...   
2016-04-03    0.039425
2016-04-04    0.037194
2016-04-05    0.011965
2016-04-06    0.003310
2016-04-07    0.001299
Name: ad_created, Length: 76, dtype: float64

In [147]:
autos['last_seen'].str[:10].value_counts(normalize=True, dropna=False).sort_index()

2016-03-05    0.001128
2016-03-06    0.003800
2016-03-07    0.004781
2016-03-08    0.006522
2016-03-09    0.009121
2016-03-10    0.010126
2016-03-11    0.011622
2016-03-12    0.022753
2016-03-13    0.008581
2016-03-14    0.012038
2016-03-15    0.015103
2016-03-16    0.015618
2016-03-17    0.026700
2016-03-18    0.007478
2016-03-19    0.014932
2016-03-20    0.019811
2016-03-21    0.020301
2016-03-22    0.020988
2016-03-23    0.018266
2016-03-24    0.018928
2016-03-25    0.018143
2016-03-26    0.015961
2016-03-27    0.014858
2016-03-28    0.019786
2016-03-29    0.021061
2016-03-30    0.023537
2016-03-31    0.022998
2016-04-01    0.022998
2016-04-02    0.025425
2016-04-03    0.024567
2016-04-04    0.023537
2016-04-05    0.129554
2016-04-06    0.231550
2016-04-07    0.137425
Name: last_seen, dtype: float64

### Checking the Regsitration Year Column

In [148]:
autos["registration_year"].describe()

count    40786.000000
mean      2005.389864
std         84.381634
min       1000.000000
25%       2000.000000
50%       2005.000000
75%       2009.000000
max       9999.000000
Name: registration_year, dtype: float64

In [149]:
autos['registration_year'].value_counts().sort_index(ascending = False).head(25)

9999       2
9000       1
8888       1
6200       1
5911       1
5000       2
4500       1
4100       1
2800       1
2019       1
2018     435
2017    1113
2016     733
2015     362
2014     649
2013     795
2012    1306
2011    1617
2010    1585
2009    2076
2008    2205
2007    2261
2006    2660
2005    2816
2004    2645
Name: registration_year, dtype: int64

In [150]:
autos['registration_year'].value_counts().sort_index(ascending = False).tail(25)

1962     4
1961     6
1960    19
1959     6
1958     4
1957     2
1956     4
1955     2
1954     2
1953     1
1952     1
1951     2
1950     1
1948     1
1943     1
1941     2
1939     1
1938     1
1937     4
1934     2
1931     1
1929     1
1927     1
1001     1
1000     1
Name: registration_year, dtype: int64

We see that the years after 2016 and before 1927 are incorrectly entered and hence we remove them

In [151]:
autos = autos[autos["registration_year"].between(1927,2016)]

In [152]:
autos['registration_year'].value_counts(normalize=True, dropna=False, bins = 20).sort_index(ascending = False).round(3)

(2011.55, 2016.0]     0.098
(2007.1, 2011.55]     0.191
(2002.65, 2007.1]     0.330
(1998.2, 2002.65]     0.225
(1993.75, 1998.2]     0.104
(1989.3, 1993.75]     0.024
(1984.85, 1989.3]     0.011
(1980.4, 1984.85]     0.004
(1975.95, 1980.4]     0.005
(1971.5, 1975.95]     0.002
(1967.05, 1971.5]     0.003
(1962.6, 1967.05]     0.002
(1958.15, 1962.6]     0.001
(1953.7, 1958.15]     0.000
(1949.25, 1953.7]     0.000
(1944.8, 1949.25]     0.000
(1940.35, 1944.8]     0.000
(1935.9, 1940.35]     0.000
(1931.45, 1935.9]     0.000
(1926.91, 1931.45]    0.000
Name: registration_year, dtype: float64

This is the distribution of the dates of registration by groups, in which we can see how the period between 1986 and 201 is the one with the highest number of registrations. 

### Exploring cars by Brand 

In [153]:
autos["brand"].unique()

array(['peugeot', 'bmw', 'volkswagen', 'smart', 'ford', 'chrysler',
       'audi', 'renault', 'sonstige_autos', 'mazda', 'porsche', 'mini',
       'mercedes_benz', 'seat', 'toyota', 'opel', 'dacia', 'nissan',
       'jeep', 'saab', 'volvo', 'jaguar', 'skoda', 'subaru', 'fiat',
       'mitsubishi', 'chevrolet', 'kia', 'hyundai', 'honda', 'citroen',
       'suzuki', 'trabant', 'daewoo', 'land_rover', 'alfa_romeo', 'lada',
       'rover', 'daihatsu', 'lancia'], dtype=object)

In [154]:
autos["brand"].value_counts()

volkswagen        8249
bmw               4761
mercedes_benz     4258
audi              3724
opel              3656
ford              2398
renault           1559
peugeot           1105
fiat               868
skoda              721
seat               701
smart              635
mazda              559
toyota             558
citroen            555
nissan             554
hyundai            413
mini               406
sonstige_autos     399
volvo              358
kia                296
honda              294
mitsubishi         281
porsche            278
alfa_romeo         255
chevrolet          252
suzuki             232
chrysler           131
dacia              122
jeep               105
land_rover          98
daihatsu            77
subaru              73
jaguar              69
saab                57
daewoo              42
trabant             36
rover               33
lancia              31
lada                25
Name: brand, dtype: int64

In [155]:
#Selecting and analyzing the top 10 brands
top_10_brands = autos["brand"].value_counts().index[:10]
top_10_brands

Index(['volkswagen', 'bmw', 'mercedes_benz', 'audi', 'opel', 'ford', 'renault',
       'peugeot', 'fiat', 'skoda'],
      dtype='object')

In [163]:
# finding out what is the mean price of the cars sold for these brands
brands_price = {}

for i in top_10_brands:
    sel = autos[autos["brand"] == i]
    brands_price[i] = sel["price_in_dollars"].mean().round()
    
sorted(brands_price.items(),key = lambda x: x[1], reverse=True)
    

[('audi', 10088.0),
 ('mercedes_benz', 9094.0),
 ('bmw', 8952.0),
 ('skoda', 6739.0),
 ('volkswagen', 6364.0),
 ('ford', 4939.0),
 ('opel', 3911.0),
 ('peugeot', 3772.0),
 ('fiat', 3710.0),
 ('renault', 3297.0)]

In [171]:
#Similarly we find out the mean number of kilometeres run for these brands
brands_km = {}

for i in top_10_brands:
    sel = autos[autos["brand"] == i]
    brands_km[i] = sel["odometer_km"].mean().round()
    
sorted(brands_km.items(), key=lambda x: x[1], reverse=True)

[('bmw', 132200.0),
 ('mercedes_benz', 130461.0),
 ('audi', 128012.0),
 ('volkswagen', 126561.0),
 ('opel', 125495.0),
 ('peugeot', 123362.0),
 ('renault', 123307.0),
 ('ford', 121253.0),
 ('fiat', 110478.0),
 ('skoda', 110430.0)]

In [172]:
combined={}

x = "Mean price ${:.2f} and {:.2f} mean km"

for i in brands_price:
    mean_price = brands_price[i]
    mean_km = brands_km[i]
    combined[i] = x.format(mean_price, mean_km)
    
combined    
    

{'volkswagen': 'Mean price $6364.00 and 126561.00 mean km',
 'bmw': 'Mean price $8952.00 and 132200.00 mean km',
 'mercedes_benz': 'Mean price $9094.00 and 130461.00 mean km',
 'audi': 'Mean price $10088.00 and 128012.00 mean km',
 'opel': 'Mean price $3911.00 and 125495.00 mean km',
 'ford': 'Mean price $4939.00 and 121253.00 mean km',
 'renault': 'Mean price $3297.00 and 123307.00 mean km',
 'peugeot': 'Mean price $3772.00 and 123362.00 mean km',
 'fiat': 'Mean price $3710.00 and 110478.00 mean km',
 'skoda': 'Mean price $6739.00 and 110430.00 mean km'}

## Storing Aggregate Data in Dataframe

In [175]:
brands_price_series = pd.Series(brands_price)
brands_price_series

volkswagen        6364.0
bmw               8952.0
mercedes_benz     9094.0
audi             10088.0
opel              3911.0
ford              4939.0
renault           3297.0
peugeot           3772.0
fiat              3710.0
skoda             6739.0
dtype: float64

In [177]:
brands_km_series = pd.Series(brands_km)
brands_km_series

volkswagen       126561.0
bmw              132200.0
mercedes_benz    130461.0
audi             128012.0
opel             125495.0
ford             121253.0
renault          123307.0
peugeot          123362.0
fiat             110478.0
skoda            110430.0
dtype: float64

In [180]:
frame = {"brand_mean_price":brands_price_series,
         "brand_mean_km":brands_km_series}

output = pd.DataFrame(frame)
output

brand_mean_price  brand_mean_km
volkswagen               6364.0       126561.0
bmw                      8952.0       132200.0
mercedes_benz            9094.0       130461.0
audi                    10088.0       128012.0
opel                     3911.0       125495.0
ford                     4939.0       121253.0
renault                  3297.0       123307.0
peugeot                  3772.0       123362.0
fiat                     3710.0       110478.0
skoda                    6739.0       110430.0

## Conclusion

As for this dataset, we first started cleaning the data and went onto to explore the various columns.
We concentrated on some columns like `brand`, `price` and `mileage` to check for any correlation they might have. So we selected the top 10 brands and calculated the mean price and milegae. 

- We see that there is a slight trent to the more expensive cars having more milegae but the difference is not too huge!

